lib/maths/CSeasonalComponentAdaptiveBucketing.cc

```diff
@@ -179,21 +179,32 @@ void CSeasonalComponentAdaptiveBucketing::shiftLevel(double shift) {
 
 void CSeasonalComponentAdaptiveBucketing::shiftSlope(core_t::TTime time, double shift) {
     if (std::fabs(shift) > 0.0) {
-        for (auto& bucket : m_Buckets) {
-            bucket.s_Regression.shiftGradient(shift);
-            bucket.s_Regression.shiftOrdinate(-shift * m_Time->regression(time));
+        const auto& centres = this->centres();
+        for (std::size_t i = 0; i < m_Buckets.size(); ++i) {
+            core_t::TTime bucketTime{time + static_cast<core_t::TTime>(centres[i] + 0.5)};
+            m_Buckets[i].s_Regression.shiftGradient(shift);
+            m_Buckets[i].s_Regression.shiftOrdinate(-shift * m_Time->regression(bucketTime));
         }
     }
 }
 
 void CSeasonalComponentAdaptiveBucketing::linearScale(core_t::TTime time, double scale) {
-    for (auto& bucket : m_Buckets) {
-        double gradientBefore{gradient(bucket.s_Regression)};
-        bucket.s_Regression.linearScale(scale);
-        double gradientAfter{gradient(bucket.s_Regression)};
-        bucket.s_Regression.shiftGradient(gradientBefore - gradientAfter);
-        bucket.s_Regression.shiftOrdinate(-(gradientBefore - gradientAfter) *
-                                          m_Time->regression(time));
+    const auto& centres = this->centres();
+    for (std::size_t i = 0; i < m_Buckets.size(); ++i) {
+        if (scale <= 1.0) {
+            m_Buckets[i].s_Regression.linearScale(scale);
+        } else {
+            double gradientBefore{gradient(m_Buckets[i].s_Regression)};
+            m_Buckets[i].s_Regression.linearScale(scale);
+            double gradientAfter{gradient(m_Buckets[i].s_Regression)};
+
+            core_t::TTime bucketTime{time + static_cast<core_t::TTime>(centres[i] + 0.5)};
+            m_Buckets[i].s_Regression.shiftGradient(gradientBefore - gradientAfter);
+            m_Buckets[i].s_Regression.shiftOrdinate(
+                -(gradientBefore - gradientAfter) * m_Time->regression(bucketTime));
+        }
+        m_Buckets[i].s_Variance += CTools::truncate(0.02 * (scale - 1.0), 0.0, 0.1) *
```

Comment:
This formula seems to contain some magic numbers.  It might be clearer if they were symbolic constants or if there was a comment explaining the reasoning for them.

Reply:
The functional form isn't magic, but the numbers are somewhat. In practice anomaly detection is not super sensitive to the exact choice of numbers. These were chosen as the smallest correction needed to essentially avoid spurious anomalies after applying a change point in some representative data sets. I think the key idea is to set the scale based on prediction^2. This means we effectively stop transient anomalies associated with small **relative** errors in our predictions. I've added a comment to this effect.

Reply:
See [51019a6](https://github.com/elastic/ml-cpp/pull/1646/commits/51019a6da2e720e304bb9dad4b400c53ab655b0f).

Reply:
I changed my mind and just reverted the non-essential parts of this PR. I think it makes sense to keep this to purely addressing causes of instability and any changes of behaviour can be made in a separate PR.